--- a/rdesktop.c
+++ b/rdesktop.c
@@ -850,6 +850,12 @@
 	if (seamless_slave)
 	{
 		seamless_socket_send(master_socket, argv[optind]);
+
+        // fake startup notification
+        system("xsetroot -cursor_name watch");
+        sleep(10);
+        system("xsetroot -cursor_name left_ptr");
+
 		return 0;
 	}
 
--- a/xwin.c
+++ b/xwin.c
@@ -3807,6 +3807,12 @@
 		XSetWMHints(g_display, sw->wnd, wmhints);
 		XFree(wmhints);
 	}
+
+    // Fake startup notification
+    if (parent == 0x00000000)
+    {
+        system("xsetroot -cursor_name left_ptr");
+    }
 }
 
 
