--- a/panel/panel.c
+++ b/panel/panel.c
@@ -633,15 +633,18 @@
 
     g_signal_connect (mi, "activate", G_CALLBACK (panel_app_quit), NULL);
     
-    mi = gtk_image_menu_item_new_with_label (_("Restart"));
-    gtk_widget_show (mi);
-    gtk_menu_shell_append (GTK_MENU_SHELL (menu), mi);
-    
-    img = gtk_image_new_from_stock (GTK_STOCK_REFRESH, GTK_ICON_SIZE_MENU);
-    gtk_widget_show (img);
-    gtk_image_menu_item_set_image (GTK_IMAGE_MENU_ITEM (mi), img);
+    if (xfce_allow_panel_customization ())
+    {
+        mi = gtk_image_menu_item_new_with_label (_("Restart"));
+        gtk_widget_show (mi);
+        gtk_menu_shell_append (GTK_MENU_SHELL (menu), mi);
+    
+        img = gtk_image_new_from_stock (GTK_STOCK_REFRESH, GTK_ICON_SIZE_MENU);
+        gtk_widget_show (img);
+        gtk_image_menu_item_set_image (GTK_IMAGE_MENU_ITEM (mi), img);
 
-    g_signal_connect (mi, "activate", G_CALLBACK (panel_app_restart), NULL);
+        g_signal_connect (mi, "activate", G_CALLBACK (panel_app_restart), NULL);
+    }
 
     mi = gtk_separator_menu_item_new ();
     gtk_widget_show (mi);
