Index: Xvnc/programs/Xserver/hw/vnc/rfbserver.c
===================================================================
--- Xvnc/programs/Xserver/hw/vnc/rfbserver.c	(revision 30)
+++ Xvnc/programs/Xserver/hw/vnc/rfbserver.c	(working copy)
@@ -650,7 +650,7 @@
         cl->cachingVersion = min(rfbCachingVersion_default, ci->cacheType);
         cl->cachingMaximumEntries = min(rfbCachingMaximumEntries_default, Swap16IfLE(ci->maximumEntries));
         cl->cachingMaintenanceAlgorithm = ci->maintenanceAlgorithm;
-        cl->cachingMinimumCacheDatasize = min(rfbCachingMinimumCacheDatasize_default, Swap16IfLE(ci->minimumCacheDataSize));
+        cl->cachingMinimumCacheDatasize = Swap16IfLE(ci->minimumCacheDataSize);
         rfbLog("Caching is on:\n");
         rfbLog("  version: %d.%d\n", cl->cachingVersion>>4, cl->cachingVersion & 0x0F);
         rfbLog("  maximum entries: %d\n", cl->cachingMaximumEntries);
